---
title: "Myocardin splicing: differential exon usage analysis"
author:
   name: "Dmytro Kryvokhyzha"
   email: dmytro.kryvokhyzha@med.lu.se
   affiliation: LUDC Bioinformatics Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:    
    toc: true
---
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, eval=FALSE)
knitr::opts_knit$set(root.dir = '/home/dmytro/Science/myocardin_splicing_Ola_Karl/')
```

## Dependencies

```{r, message=FALSE, eval=TRUE}
library(DEXSeq)
library(BiocParallel)
library(RColorBrewer)
library(ggplot2)
library(sm)
library(data.table)
library(gridExtra)
```

## Data

```{r, message= FALSE, echo=FALSE, eval=TRUE}
rdataFile <- 'intermediate/DEXSeq.RData'
if (file.exists(rdataFile)){
  load(rdataFile)
  knitr::opts_chunk$set(eval=FALSE)
}
```

Load the counts:

```{r, message=FALSE}
countFiles <- list.files('results/tables/DEXSeq', pattern="^CMV*|^MYO*",
                         full.names=TRUE)
sampleTable <- data.frame(row.names = c("CMV1", "CMV2", "CMV4", "CMV5",
                                        "MYO1", "MYO2", "MYO4", "MYO5"),
                          condition = c(rep('control', 4), rep('treatment', 4)))
flattenedFile <- 'intermediate/DEXSeq/DEXSeq_symbol.gtf'
dxd <- DEXSeqDataSetFromHTSeq(countFiles,
                              sampleData=sampleTable,
                              design= ~ sample + exon + condition:exon,
                              flattenedfile=flattenedFile)
#colData(dxd)
```

```{r, echo=FALSE, eval=FALSE}
# Subset data for development
dxd <- dxd[sample(nrow(dxd), 10000), ]
```

Remove lowly expressed

```{r}
#hist(rowSums(featureCounts(dxd)), breaks = 10000000, xlim=c(0, 30))
dim(dxd)[1]
keep <- which(rowSums(featureCounts(dxd))>0)
length(keep)
dxd <- dxd[keep,]
```

Print the first 3 lines of the feature data annotation:

```{r, eval=TRUE}
knitr::kable(head(rowRanges(dxd), 3))
```

Check the design:

```{r, eval=TRUE}
knitr::kable(sampleAnnotation(dxd))
```

Exon counts are split into two categories. 
The first 8 columns correspond to the number of reads mapping to 
out exonic regions and the last 8 columns correspond to the sum of the counts 
mapping to the rest of the exons from the same gene on each sample.

```{r}
head(counts(dxd))
split(seq_len(ncol(dxd)), colData(dxd)$exon )
```

## Library size & counts distribution

This is rather **technical check** and **should not be used** in any interpretation.

DEXSeq gives very high numbers because reads that touch multiple counting bins provide evidence for the presence of each of the bins, therefore they are counted for each of the bins. So, the evidence is not independent. The sum of bins counts is typically larger than the gene count.

The dependence is not a problem since the testing in DEXSeq is marginal (bin by bin). Therefore 

**Library size barplot & counts distribution boxplot**

```{r, fig.height=7, eval=TRUE}
colors2 <- brewer.pal(3, 'Accent')[c(1,2)]
condition_color <- c(rep(colors2[1], 4),rep(colors2[2], 4))
par(mfcol=c(2,1))
# Library size
read_counts <- featureCounts(dxd)
libsize <- colSums(read_counts)/1000000
barplot(libsize,
        ylim = c(0,80),
        names.arg=names(read_counts),
        col= condition_color,
        ylab = 'Million reads',
        main = 'Library size',
        las =2)
abline(h=median(libsize), lty=2)
# counts distribution b oxplot
logCPM <- log2((read_counts+1)/colSums(read_counts+1)*1e+06)
boxplot(logCPM,
        xlab='',
        ylab='log2(CPM)',
        notch	=T,
        col = condition_color,
        outline=FALSE,
        main = 'Counts distribution',
        las=2)
abline(h=median(as.matrix(logCPM)), lty=2, col='lightgrey')
```

**Samples density plot**

```{r, eval=TRUE}
condition <- c(rep('control', 4), rep('treatment', 4))
logCPM_dens <- data.frame(logCPM=as.vector(t(logCPM)),
                          samples=rep(colnames(logCPM), dim(logCPM)[1]),
                          condition=rep(condition, dim(logCPM)[1]))
# samples density plot
sm.density.compare(logCPM_dens$logCPM,
                   logCPM_dens$samples,
                   xlab='Log2 counts per million')
title(main='Samples counts distribution')
colfill<-c(2:(2+length(levels(logCPM_dens$samples)))) 
legend('topright',
       levels(logCPM_dens$samples),
       col=colfill,
       lty=c(1:length(logCPM_dens$samples)),
       lwd=1,
       ncol=2,
       cex=0.7)
```

Condition density plot

```{r, eval=TRUE}
ggplot(logCPM_dens, aes(x = logCPM, colour = condition)) +
  geom_density(alpha = 0.5, size = 1.5) + theme_bw() +
  theme(legend.position = 'top') + xlab(expression(log[2](count + 1)))
```


## Normalization & dispersion estimation

The same method as in DESeq2:

```{r}
dxd <- estimateSizeFactors(dxd)
dxd <- estimateDispersions(dxd)
```

Plot dispersion:

```{r, eval=TRUE}
plotDispEsts(dxd)
```

## Differential exon usage

### GLM

*DEXSeq* fits a generalized linear model with the formula 
`~ sample + exon + condition:exon` and compare it to the null model) `~ sample + exon`.
The deviances of both fits are compared using a Ï‡2-distribution, providing a p value.

We test whether the fraction of the genes reads that fall onto the exon under the test 
differs significantly between the experimental conditions.

```{r}
dxd <- testForDEU(dxd)
dxd <- estimateExonFoldChanges(dxd)
```

### Results

```{r}
dxr <- DEXSeqResults(dxd)
knitr::kable(head(dxr))
```

```{r, eval=FALSE}
save.image(file = "intermediate/DEXSeq.RData")
```

Description of each column:

```{r, eval=TRUE}
mcols(dxr)$description
```

### Exon level

How many exonic exons are significant with a false discovery rate of 1%?

```{r, eval=TRUE}
sum(dxr$padj < 0.01, na.rm = T)
```

```{r, eval=TRUE}
plotMA(dxr, cex=1)
```

### Gene level

How many genes are affected?

```{r, eval=TRUE}
sum(tapply(dxr$padj < 0.01, dxr$groupID, any), na.rm = T)
```

DEXSeq returns one p-value for each exonic region.
However, sometimes it is informative to know what is the FDR at the gene level,
i.e. knowing the number of genes with at least one differentially used exon
while keeping control of FDR.

```{r, eval=TRUE}
dxr_geneFDR <- perGeneQValue(dxr)
dxr_geneFDR <- as.data.frame(dxr_geneFDR[order(dxr_geneFDR)])
colnames(dxr_geneFDR) <- 'gene_level_FDR'
sum(dxr_geneFDR$gene_level_FDR < 0.01)
```

### Visualization

Test with **MYOCD**.

```{r, eval=TRUE}
head(dxr[order(dxr$padj),'groupID'])
```

```{r, eval=TRUE}
plotDEXSeq(dxr, 'ENSG00000141052.18_MYOCD',
           displayTranscripts=TRUE,
           splicing=TRUE,
           legend=TRUE,
           cex.axis=1.2, cex=1.3, lwd=2 )
```

CHRM3

```{r, eval=TRUE}
dxrDF <- as.data.frame(dxr)
dxrDF$padj <- p.adjust(dxrDF$pvalue, method = 'fdr')
#sum(p.adjust(dxrDF$pvalue, method = 'fdr') < 0.01, na.rm = T)
CHRM3 <- dxrDF[grep('ENSG00000133019.12', rownames(dxrDF)), ]
```

## Control genes

*MYH11*, *MCAM* and *SYNM* can be used to verify that *MYOCD* over-expression worked

There are also much less responsive target genes: *RRAS*, *RBPMS*, and *RBPMS2*.
  
We also have qPCR information in *MYH11, MCAM, KCNMB1*. 

Differentially splicing of the following genes is of interest:
*ACTN1, SMTN, NFIC, BICD2, TEAD1, TPM1, TPM2, PDLIM7, FHL1, CSR2,
TGFB11, PKD1, MYL6, PDLIM5, FLNA, MBNL1, ZFR, FNBP1, SORBS1, PIP5K1C,
TNRC18, SVIL, KLHL42, LARGE1, MICAL3*

```{r, eval=TRUE}
interesting_genes <- c('MYOCD', 'MYH11', 'MCAM', 'SYNM', 'RRAS', 'RBPMS', 'RBPMS2', 'KCNMB1',
                       'ACTN1', 'SMTN', 'NFIC', 'BICD2', 'TEAD1', 'TPM1', 'TPM2',
                       'PDLIM7', 'FHL1', 'CSR2', 'TGFB11', 'PKD1', 'MYL6', 'PDLIM5',
                       'FLNA', 'MBNL1', 'ZFR', 'FNBP1', 'SORBS1', 'PIP5K1C',
                       'TNRC18', 'SVIL', 'KLHL42', 'LARGE1', 'MICAL3')
```

Extract the results for these genes: 

```{r, eval=TRUE}
interesting_genes_rows <- grep(paste(interesting_genes,collapse="|"), dxr$groupID)
dxr_interesting_genes <- dxr[interesting_genes_rows,]
knitr::kable(head(dxr_interesting_genes))
```

How many genes are significant?

FDR < 0.01

```{r, eval=TRUE}
sum(tapply(dxr_interesting_genes$padj < 0.01, dxr_interesting_genes$groupID, any), na.rm = T)
```

FDR < 0.05

```{r, eval=TRUE}
sum(tapply(dxr_interesting_genes$padj < 0.05, dxr_interesting_genes$groupID, any), na.rm = T)
```

What are those genes?

```{r, eval=TRUE}
unique(dxr_interesting_genes[which(dxr_interesting_genes$padj < 0.05), 'groupID'])
```

## Select cadidates

Look at top percentiles

```{r, eval=TRUE}
colors5 <- brewer.pal(5, 'Accent')
topPerc <- c(0.05, 0.01, 0.001, 0.0001)
topPerc_log2FC <- quantile(dxr$log2fold_treatment_control, 1-topPerc/2, na.rm=T)
topPerc_expr <- quantile(c(dxr$treatment, dxr$control), 1-topPerc, na.rm=T)
round(topPerc_log2FC, 0)
round(topPerc_expr, 0)
```

Plot each stats and define tails to cut-off thresholds:

```{r}
pdf('results/figures/padj_logFC_expr.pdf', width = 8, height = 5)
ggplot(as.data.frame(dxr), aes(x=padj)) +
      geom_density() + theme_bw() +
      geom_vline(xintercept = topPerc[1], alpha=0.5, color = colors5[1], size=1) +
      geom_vline(xintercept = topPerc[2], alpha=0.5, color = "blue", size=1) +
      geom_vline(xintercept = topPerc[3], alpha=0.5, color = colors5[3], size=1) +
      geom_vline(xintercept = topPerc[4], alpha=0.5, color = colors5[4], size=1)
ggplot(as.data.frame(dxr), aes(x=log2fold_treatment_control)) +
      geom_density() + theme_bw() +
      geom_vline(xintercept = topPerc_log2FC[1], alpha=0.5, color = colors5[1], size=1) +
      geom_vline(xintercept = topPerc_log2FC[2], alpha=0.5, color = colors5[2], size=1) +
      geom_vline(xintercept = topPerc_log2FC[3], alpha=0.5, color = colors5[3], size=1) +
      geom_vline(xintercept = topPerc_log2FC[4], alpha=0.5, color = colors5[4], size=1) +
      geom_vline(xintercept = -topPerc_log2FC[1], alpha=0.5, color = colors5[1], size=1) +
      geom_vline(xintercept = -topPerc_log2FC[2], alpha=0.5, color = colors5[2], size=1) +
      geom_vline(xintercept = -topPerc_log2FC[3], alpha=0.5, color = colors5[3], size=1) +
      geom_vline(xintercept = -topPerc_log2FC[4], alpha=0.5, color = colors5[4], size=1) +
      geom_vline(xintercept = 1, color = "blue", alpha=0.5, size=1) +
      geom_vline(xintercept = -1, color = "blue", alpha=0.5, size=1)
expr <- melt(as.matrix(dxr[, c('treatment','control')]))
colnames(expr)[c(2,3)] <- c('condition', 'expression')
ggplot(expr, aes(x=expression)) +
      geom_density() + theme_bw() +
      geom_vline(xintercept = topPerc_expr[1], alpha=0.5, color = colors5[1], size=1) +
      geom_vline(xintercept = topPerc_expr[2], alpha=0.5, color = colors5[2], size=1) +
      geom_vline(xintercept = topPerc_expr[3], alpha=0.5, color = colors5[3], size=1) +
      geom_vline(xintercept = topPerc_expr[4], alpha=0.5, color = colors5[4], size=1) +
      geom_vline(xintercept = 10, color = "blue", alpha=0.5, size=1)
dev.off()
```

Filter:

```{r, eval=TRUE}
threshold_padj <- 0.01
threshold_log2FC <- 1
threshold_expr <- 10
# I also filter for gene level FDR

dxr_top <- as.data.frame(dxr)
dxr_top <- dxr_top[which(dxr_top$padj < threshold_padj &
                  c(dxr_top$log2fold_treatment_control < -threshold_log2FC |
                    dxr_top$log2fold_treatment_control > threshold_log2FC)   & 
                  c(dxr_top$control > threshold_expr | dxr_top$treatment > threshold_expr) &
                    dxr_top$groupID %in% rownames(dxr_geneFDR)),]
length(unique(dxr_top$groupID))
```

Check how many of the candidate genes are in this list:

```{r, eval=TRUE}
dxr_top_candidates_rows <- grep(paste(interesting_genes,collapse="|"), dxr_top$groupID)
dxr_top_candidates <- dxr_top[dxr_top_candidates_rows,]
unique(dxr_top_candidates$groupID)
```

## Write Results

### HTML Report

The results for of FDR<0.05 are exported to an HTML report:

```{r}
filePath <- 'results/reports/DEXSeqReport/'
dir.create(filePath, showWarnings = FALSE)

DEXSeqHTML(dxr, FDR=0.05, color=colors2,
           path=filePath,
           file='DEXSeq_results.html',
           BPPARAM=ncores)
```

### Exons

```{r}
write.table(dxr,'results/tables/DEXSeq/DEXSeq_results.csv',
            row.names = F, sep = '\t')
write.table(dxr_top,'results/tables/DEXSeq/DEXSeq_results_top_candidates.csv',
            row.names = F, sep = '\t')
```

### Gene-level FDR

Write the gene-level FDR values to a file:

```{r}
write.table(dxr_geneFDR,
            'results/tables/DEXSeq/DEXSeq_results_geneFDR.csv',
            row.names = T, sep = '\t')
```

- Report: `results/reports/DEXSeqReport/testForDEU.html`.

- Figures that can be used in a publication: `results/reports/DEXSeqReport/files/*.svg`

- Results of the differential exon usage analysis: `results/tables/DEXSeq/DEXSeq_results.csv`

- Gene-level FDR values: `results/tables/DEXSeq/DEXSeq_results_geneFDR.csv`

### IGV input

```{r}
igv_all <- as.data.frame(dxr[,c('genomicData', 'log2fold_treatment_control',
                                'treatment', 'control', 'padj')])
igv_all$genomicData.width <- NULL
igv_all$genomicData.strand <- NULL
colnames(igv_all)[c(1:4)] <- c('CHR', 'START', 'END', 'log2FC')
igv_all$ID <- row.names(igv_all)

igv <- igv_all[,c('CHR', 'START', 'END', 'ID', 'log2FC', 'treatment', 'control', 'padj')]
igv <- igv[!is.na(igv$log2FC),]
igv$`-log(padj)` <- -log(igv$padj)
igv$padj <- NULL
igv <- igv[order(igv$CHR, igv$START),]
write.table(igv,
            'intermediate/IGV/DEXSeq_results_IGV.igv',
            row.names = F,
            sep = '\t',
            quote = F)
```


## Preprocessing using featureCounts

See https://github.com/vivekbhr/Subread_to_DEXSeq.

It seems like there is no benefits of using featureCounts with DEXSeq.

```{bash, include=FALSE}
cd results/tables/DEXSeq/
for i in DEXSeq_results*.csv;
  do
    soffice --headless \
            --convert-to xlsx:"Calc MS Excel 2007 XML" \
            --infilter="csv:9,34,UTF8" $i;
  done
cd -
```
