---
title: "Myocardin splicing: DEXSeq analysis of the MYOCD groups in the MSAT"
author:
   name: "Dmytro Kryvokhyzha"
   email: dmytro.kryvokhyzha@med.lu.se
   affiliation: LUDC Bioinformatics Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: show
    toc: true
---
    
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, eval=TRUE)
```

I split the MSAT muscle expression samples into two groups based on the MYOCD expression
and tested of there is differential exon usage between them.

## Dependencies

```{r, message=FALSE, eval=TRUE}
library(DEXSeq)
library(RColorBrewer)
```

## Data

Load meta:

```{r}
meta <- read.table('data/MSAT_samples_RNA_seq.csv',
                   header = T)
row.names(meta) <- meta$ID_Nr_RNA_Seq
```

Load the gene data:

```{r, message=FALSE}
g <- read.table(gzfile('data/MSAT_all_raw_tpm_mean.csv.gz'), header=T, check.names=F)
rownames(g) <- g$gene
```

<!-- Get DEXSeq bin-to-exon info: -->

<!-- ```{bash, eval = F} -->
<!-- mkdir -p intermediate/MSAT_correlation_MYOCD/ -->
<!-- grep -E "ENSG00000072110|DEXSeq_groupID" results/tables/DEXSeq/DEXSeq_results_bins2exons.csv \ -->
<!--   > intermediate/MSAT_correlation_MYOCD/ACTN1_bins2exons.csv -->
<!-- grep -E "ENSG00000152601|DEXSeq_groupID" results/tables/DEXSeq/DEXSeq_results_bins2exons.csv \ -->
<!--   > intermediate/MSAT_correlation_MYOCD/MBNL1_bins2exons.csv -->
<!-- grep -E "ENSG00000035403|DEXSeq_groupID" results/tables/DEXSeq/DEXSeq_results_bins2exons.csv \ -->
<!--   > intermediate/MSAT_correlation_MYOCD/VCL_bins2exons.csv -->
<!-- grep -E "ENSG00000058272|DEXSeq_groupID" results/tables/DEXSeq/DEXSeq_results_bins2exons.csv \ -->
<!--   > intermediate/MSAT_correlation_MYOCD/PPP1R12A_bins2exons.csv -->
<!-- ``` -->

<!-- ## ACTN1 - ENSG00000072110 -->

<!-- ACTN1 shows DEU in E012, E014, E015, E016, E017, E019. I checked correlation for these exons. -->

<!-- Find exons: -->

<!-- ```{bash} -->
<!-- grep -E "E012|E014|E015|E016|E017|E019" intermediate/MSAT_correlation_MYOCD/ACTN1_bins2exons.csv | cut -f 8 | sort -u | sed 's/\..*//g;s/"//g' -->
<!-- ``` -->

<!-- ```{r, message=F} -->
<!-- exons <- c('ENSE00001518409', 'ENSE00002324197', 'ENSE00002456450', -->
<!--            'ENSE00002462848', 'ENSE00002470358', 'ENSE00003470414', -->
<!--            'ENSE00003626569', 'ENSE00003643367', 'ENSE00003675586') -->
<!-- ``` -->

<!-- `ENSE00002456450` overlaps both E012 and E014 that are differently regulated. -->
<!--   - E012 is up-regulated -->
<!--   - E014 is down-regulated by MYOCD. -->

<!-- `ENSE00002456450` overlaps only E019 that is up-regulated by MYOCD. -->

## MYOCD

Extract the MYOCD gene:

```{r}
myocd <- g[g$gene=="ENSG00000141052", as.character(meta$ID_Nr_RNA_Seq[meta$RNA_Type=="Muscle"])]
LOGmyocd <- log(myocd)
q <- .30
myocdQ <- quantile(LOGmyocd, c(0+q, 1-q))
```

Check the distribution

```{r}
hist(as.numeric(LOGmyocd), breaks = 30)
abline(v=myocdQ, col='red', lwd=2)
```

Split into two groups by selecting samples in top and down 0.3 quantile of MYCD expression:

```{r}
down <- which(LOGmyocd<=myocdQ[[1]])
up <- which(LOGmyocd>=myocdQ[[2]])
myocdD <- data.frame(samples = c(colnames(myocd)[down], colnames(myocd)[up]),
                     expression = c(rep('down', length(down)), rep('up', length(up))))
write.table(myocdD,
            'intermediate/MSAT_correlation_MYOCD/MOYCD_expression_groups.csv',
            row.names = F,
            quote = F,
            sep = '\t')
```

## DEXSeq

Load the counts:

```{r}
countFiles <- paste('results/tables/DEXSeq_MSAT/', myocdD$samples, '_bins.csv', sep = "")

flattenedFile <- 'intermediate/DEXSeq/DEXSeq_symbol.gtf'

formulaFullModel <- formula(~ sample + exon + expression:exon)
formulaReducedModel <- formula(~ sample + exon)

dxd <- DEXSeqDataSetFromHTSeq(countFiles,
                              sampleData=myocdD,
                              design= formulaFullModel,
                              flattenedfile=flattenedFile)
#colData(dxd)
```

genes of interest:

```{r}
genes_interest <- c('ENSG00000072110', 'ENSG00000152601', 'ENSG00000035403', 'ENSG00000058272')
```

Remove lowly expressed but keep all genes of interest

```{r}
#hist(rowSums(featureCounts(dxd)), breaks = 10000000, xlim=c(0, 30))
dim(dxd)[1]
dxd_genes_interest <- grep(paste(genes_interest, collapse = "|"), rownames(featureCounts(dxd)))
keep <- which(rowSums(featureCounts(dxd)!=0)>=26)
#length(keep)
dxd <- dxd[union(dxd_genes_interest, keep),]
dim(dxd)[1]
```

Estimate size factors

```{r}
dxd <- estimateSizeFactors(dxd)
```

Keep only genes of interest and estimate the dispersion:

```{r}
dxd_genes <- dxd[grep(paste(genes_interest, collapse = "|"), rownames(featureCounts(dxd))), ]

# ncores <- MulticoreParam(8)
dxd_genes <- estimateDispersions(dxd_genes,
                                 formula = formulaFullModel) #,
                                 # BPPARAM=ncores)
```

Plot dispersion:

```{r, eval=TRUE}
plotDispEsts(dxd_genes)
```

## Results

Test DEU:

```{r}
dxd_genes <- testForDEU(dxd_genes,
                        reducedModel = formulaReducedModel,
                        fullModel = formulaFullModel)
dxd_genes <- estimateExonFoldChanges(dxd_genes,
                                      fitExpToVar='expression')
```

Check the results:

```{r}
dxr <- DEXSeqResults(dxd_genes)
dxr_genes_interest <- dxr[grep(paste(genes_interest, collapse = "|"), rownames(dxr)),]
knitr::kable(dxr_genes_interest)
```

Exon level

```{r, eval=TRUE}
sum(dxr$padj < 0.1, na.rm = T)
```

```{r, eval=TRUE}
plotMA(dxr, cex=1)
```

Gene level

```{r, eval=TRUE}
sum(tapply(dxr$padj < 0.1, dxr$groupID, any), na.rm = T)
```

### Visualization

Test with **MYOCD**.

```{r, eval=TRUE}
head(dxr[order(dxr$padj),'groupID'])
```

```{r, eval=TRUE}
plotDEXSeq(dxr, 'ENSG00000058272.19_PPP1R12A',
           displayTranscripts=TRUE,
           splicing=TRUE,
           legend=TRUE,
           cex.axis=1.2, cex=1.3, lwd=2 )
```

## Write Results

### HTML Report

The results for FDR<0.1 are exported to an HTML report:

```{r}
filePath <- 'results/reports/DEXSeqReport_MSAT/'
dir.create(filePath, showWarnings = FALSE)

colors2 <- brewer.pal(3, 'Accent')[c(1,2)]

DEXSeqHTML(dxr, FDR=0.1, color=colors2,
           path=filePath,
           fitExpToVar="expression",
           file='DEXSeq_MSAT_results.html')
```

### Exons

```{r}
write.table(dxr,'results/tables/DEXSeq_MSAT/DEXSeq_MSAT_results.csv',
            row.names = F, sep = '\t')
```

## Correlate MSAT results




