---
title: "Myocardin splicing: overlap between DEXSeq rMATS ISA results"
author:
   name: "Dmytro Kryvokhyzha"
   email: dmytro.kryvokhyzha@med.lu.se
   affiliation: LUDC Bioinformatics Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: show
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, eval=TRUE)
knitr::opts_knit$set(root.dir = '../')
```

## Dependecies

```{r, message=FALSE, eval=TRUE}
library(DEXSeq)
library(GenomicFeatures)
library(GenomicRanges)
library(VennDiagram)
library(RColorBrewer)
library(mygene)
dir.fig <- 'results/figures/rMATS_DEXSeq_ISA_overlap'
dir.table <- 'results/tables/rMATS_DEXSeq_ISA_overlap'
dir.create(dir.fig, showWarnings = F)
dir.create(dir.table, showWarnings = F)
```

## Functions

```{r}
getrMATS <- function(type, target){
   # type <- 'A5SS'
   # target <- 'exon_long'
   rMATS <- read.table(paste('results/tables/rMATS/rMATS_all_' , type, '.csv', sep=""),
                             header = T, sep = '\t')
   rMATS <- rMATS[which(rMATS$FDR<=0.05),]
   rMATS_coord <- data.frame(do.call('rbind', strsplit(as.character(rMATS[,target]),
                                                    '-', fixed=TRUE)))
   start <- paste(target, '_start', sep = "")
   end <- paste(target, '_end', sep = "")
   rMATS[,start] <- as.numeric(levels(rMATS_coord$X1))[rMATS_coord$X1]
   rMATS[,end] <- as.numeric(levels(rMATS_coord$X2))[rMATS_coord$X2]
   rMATS$IncLevelDifference <- rMATS$IncLevelDifference
   rMATS <- merge(rMATS, exon_id[,c('exon_chr', 'exon_start', 'exon_end',
                                      'exon_id', 'exon_number')],
                           by.x=c('Chr', start, end),
                           by.y = c('exon_chr', 'exon_start', 'exon_end'))
   rMATSexons <- GRanges(seqnames=rMATS$Chr,
                     ranges=IRanges(rMATS[,start], rMATS[,end]),
                     strand=rMATS$Strand,
                     mcols=rMATS[,c(target, 'PValue', 'FDR', 'IncLevelDifference',
                                 'exon_id')])
   rMATS[,start] <- NULL
   rMATS[,end] <- NULL
   return(list(rMATS, rMATSexons))
}

findOverlap2 <- function(rMATSd, rMATSname){
   #rMATSd <- rMATS_SEexon
   #rMATSname <- 'rMATS_SE'
   overlap <- as.data.frame(findOverlaps(exonDEXSeq, rMATSd))
   # combine hits
   exonDEXSeq_ordered <- as.data.frame(exonDEXSeq[overlap$queryHits,])
   rMATSd_ordered <- as.data.frame(rMATSd[overlap$subjectHits,])
   colnames(exonDEXSeq_ordered) <- paste('DEXSeq', sub("mcols.", "", colnames(exonDEXSeq_ordered)),
                                         sep = "_")
   colnames(exonDEXSeq_ordered)[1]<- 'DEXSeq_chr'
   colnames(rMATSd_ordered) <- paste(rMATSname, sub("mcols.", "", colnames(rMATSd_ordered)),
                                         sep = "_")
   exonDEXSeq_rMATSd <- cbind(exonDEXSeq_ordered, rMATSd_ordered)
   first_columns <- c('DEXSeq_groupID', 'DEXSeq_featureID')
   exonDEXSeq_rMATSd <- exonDEXSeq_rMATSd[,c(first_columns,
                                            setdiff(colnames(exonDEXSeq_rMATSd), first_columns))]
   exonDEXSeq_rMATSd <- exonDEXSeq_rMATSd[,c(setdiff(colnames(exonDEXSeq_rMATSd),
                                                       'DEXSeq_transcripts'), 'DEXSeq_transcripts')]
   exonDEXSeq_rMATSd <- unique(exonDEXSeq_rMATSd[order(exonDEXSeq_rMATSd$DEXSeq_chr,
                                                exonDEXSeq_rMATSd$DEXSeq_start,
                                                exonDEXSeq_rMATSd$DEXSeq_end),])
   return(exonDEXSeq_rMATSd)
}

findOverlap3 <- function(findOverlap2D){
   #findOverlap2D <- dexseq_sing_transcrips
   findOverlap2D_transcripts <- unique(unlist(strsplit(gsub('c\\(|\\)|,', '',
                                                 findOverlap2D$DEXSeq_transcripts), " ")))
   findOverlap2D_ISA_transcripts <- intersect(ISA$transcript, findOverlap2D_transcripts)
   
   # add ISA column to the results overlap
   findOverlap2D_ISA <- findOverlap2D
   findOverlap2D_ISA$ISA_overlap <- NA
   for (i in c(1:length(findOverlap2D_ISA$DEXSeq_transcripts))){
      #i <- 16
      transc_list <- unique(unlist(strsplit(gsub('c\\(|\\)|,', '',
                findOverlap2D_ISA$DEXSeq_transcripts[i]), " ")))
      transc_list_ISA <- intersect(findOverlap2D_ISA_transcripts, transc_list)
      if (!identical(transc_list_ISA, character(0))){
         findOverlap2D_ISA$ISA_overlap[i] <- paste(transc_list_ISA, collapse = ",")
      }
   }
   findOverlap2D_ISA <- findOverlap2D_ISA[,c(setdiff(colnames(findOverlap2D_ISA),
                                               'DEXSeq_transcripts'), 'DEXSeq_transcripts')]
   return(list(findOverlap2D_ISA,
               findOverlap2D_transcripts))
}

rMATS2Tx <- function(rMATSd){
   #rMATSd <- rMATS_SEexon
   rMATS_transcript <- as.data.frame(findOverlaps(transcripts, rMATSd))
   transcripts_ordered <- as.data.frame(transcripts[rMATS_transcript$queryHits,],
                                              row.names = NULL)
   rMATSd_ordered <- as.data.frame(rMATSd[rMATS_transcript$subjectHits,])
   rMATSd_ordered$transcript <- transcripts_ordered$tx_name
   rMATSd_transcript <- unique(rMATSd_ordered$transcript)
   return(rMATSd_transcript)
}
```

## Data

### Transcrips

```{r, message=FALSE}
transcriptDb <- makeTxDbFromGFF("data/reference/gencode.v35.primary_assembly.annotation.gtf",
                                format="gtf")
transcripts <- unlist(transcriptsBy(transcriptDb, "exon", use.names=TRUE))
```

### exon_id

```{r}
exon_id <- read.table('results/tables/DEXSeq/DEXSeq_results_bins2exons.csv',
                      header = T, sep = '\t')
```

### DEXSeq

```{r}
dexseq <- read.table('results/tables/DEXSeq/DEXSeq_results.csv', header = T, sep = '\t')
dexseq_sing <- dexseq[which(dexseq$padj<=0.01),]

exonDEXSeq <- GRanges(seqnames=dexseq_sing$genomicData.seqnames,
                      ranges=IRanges(dexseq_sing$genomicData.start,
                                     dexseq_sing$genomicData.end),    
                      strand=dexseq_sing$genomicData.strand,
                      mcols=dexseq_sing[,
                      c('groupID', 'featureID', 'log2fold_treatment_control', 'transcripts')])
```

How many exons the DEXSeq bins cover:

```{r}
dim(dexseq_sing)[1]
length(unique(unlist(strsplit(as.character(dexseq_sing$exon_id), ","))))
```

### rMATS

SE

```{r}
rMATS_SEall <- getrMATS('SE', 'exon_target')
rMATS_SE <- rMATS_SEall[[1]]
rMATS_SEexon <- rMATS_SEall[[2]]
```

A5SS

```{r}
rMATS_A5SSall <- getrMATS('A5SS', 'exon_long')
rMATS_A5SS <- rMATS_A5SSall[[1]]
rMATS_A5SSexon <- rMATS_A5SSall[[2]]
```

A3SS

```{r}
rMATS_A3SSall <- getrMATS('A3SS', 'exon_long')
rMATS_A3SS <- rMATS_A3SSall[[1]]
rMATS_A3SSexon <- rMATS_A3SSall[[2]]
```

MXE

```{r}
rMATS_MXE <- read.table('results/tables/rMATS/rMATS_sign_MXE.csv', header = T, sep = '\t')
rMATS_MXE_1_all <- getrMATS('MXE', 'exon_1')
rMATS_MXE_1exon <- rMATS_MXE_1_all[[2]]
rMATS_MXE_2_all <- getrMATS('MXE', 'exon_2')
rMATS_MXE_2exon <- rMATS_MXE_2_all[[2]]
```

RI

```{r}
rMATS_RIall <- getrMATS('RI', 'exon_ir')
rMATS_RI <- rMATS_RIall[[1]]
rMATS_RIexon <- rMATS_RIall[[2]]
```

### Isoform Switch Analysis

Isoform Switch Analysis (ISA)

```{r}
# ISA <- read.table('results/tables/ISA/ISA_FDR_0.05.csv', header = T, sep = '\t')
# head(ISA)
ISA <- read.table('results/tables/IsoformSwitchAnalyzeR/isoformSwitchAnalysis.csv', header = T, sep = '\t')
ISA <- ISA[ISA$padj<=0.05, ]
colnames(ISA)[1] <- 'transcript'
```

## Find exon overlap

### rMATS_SE vs DEXSeq

How many exons overlap

```{r}
DEXSeq_rMATS_SE <- findOverlap2(rMATS_SEexon, 'rMATS_SE')
DEXSeq_rMATS_SE_L <- length(unique(DEXSeq_rMATS_SE[,c('rMATS_SE_exon_target')]))
rMATS_SE_L <- dim(unique(rMATS_SE[,c("Chr", "exon_target")]))[1]

rMATS_SE_L
DEXSeq_rMATS_SE_L
```

Correlation in results

```{r}
DEXSeq_rMATS_SE_unique <- DEXSeq_rMATS_SE[
                            !duplicated(DEXSeq_rMATS_SE[,
                            c('rMATS_SE_seqnames', 'rMATS_SE_start', 'rMATS_SE_end')]),]
# correct sign
DEXSeq_rMATS_SE_unique$rMATS_SE_IncLevelDifference <-
   DEXSeq_rMATS_SE_unique$rMATS_SE_IncLevelDifference*-1

pdf('results/figures/SE_cor.pdf', width = 5, height = 4)
   plot(DEXSeq_rMATS_SE_unique$DEXSeq_log2fold_treatment_control,
        DEXSeq_rMATS_SE_unique$rMATS_SE_IncLevelDifference,
        xlab="DEXSeq log2FC", ylab="SE rMATS IncLevelDifference",
        cex=1.1, pch=16)
   abline(lm(rMATS_SE_IncLevelDifference~DEXSeq_log2fold_treatment_control,
             data=DEXSeq_rMATS_SE_unique), lty=2)
   legend('topleft', legend = paste('r = ',
          round(cor(DEXSeq_rMATS_SE_unique$DEXSeq_log2fold_treatment_control,
                    DEXSeq_rMATS_SE_unique$rMATS_SE_IncLevelDifference), 2), sep = ""),
          bty='n', bg='white', cex= 1)
dev.off()

cor.test(DEXSeq_rMATS_SE_unique$DEXSeq_log2fold_treatment_control,
                 DEXSeq_rMATS_SE_unique$rMATS_SE_IncLevelDifference, method = "spearman")
```

### rMATS_SE vs ISA

```{r}
rMATS_SE_Tx <- rMATS2Tx(rMATS_SEexon)
rMATS_SE_ISA_L <-length(intersect(rMATS_SE_Tx, ISA$transcript))

length(ISA$transcript)
length(rMATS_SE_Tx)
rMATS_SE_ISA_L
```

### rMATS_SE_DEXSeq vs ISA

```{r}
DEXSeq_rMATS_SE_ISAall <- findOverlap3(DEXSeq_rMATS_SE)
DEXSeq_rMATS_SE_ISA <- DEXSeq_rMATS_SE_ISAall[[1]]
DEXSeq_rMATS_SE_transcripts <- DEXSeq_rMATS_SE_ISAall[[2]]
DEXSeq_rMATS_SE_ISA_L <- length(unique(unlist(strsplit(as.character(
                                   DEXSeq_rMATS_SE_ISA$ISA_overlap), ","))))

length(ISA$transcript)
length(DEXSeq_rMATS_SE_transcripts)
DEXSeq_rMATS_SE_ISA_L
```

### rMATS_A5SS vs DEXSeq

How many exons overlap

```{r}
DEXSeq_rMATS_A5SS <- findOverlap2(rMATS_A5SSexon, 'rMATS_A5SS')
DEXSeq_rMATS_A5SS_L <- length(unique(DEXSeq_rMATS_A5SS[,'rMATS_A5SS_exon_long']))
rMATS_A5SS_L <- dim(unique(rMATS_A5SS[,c("Chr", "exon_long")]))[1]

rMATS_A5SS_L
DEXSeq_rMATS_A5SS_L
```

Relationship in results

```{r}
plot(DEXSeq_rMATS_A5SS$DEXSeq_log2fold_treatment_control,
     DEXSeq_rMATS_A5SS$rMATS_A5SS_IncLevelDifference,
     xlab="DEXSeq log2FC", ylab="rMATS IncLevelDifference", main="A5SS")
```

### rMATS_A5SS vs ISA

```{r}
rMATS_A5SS_Tx <- rMATS2Tx(rMATS_A5SSexon)
rMATS_A5SS_ISA_L <- length(intersect(rMATS_A5SS_Tx, ISA$transcript))

length(ISA$transcript)
length(rMATS_A5SS_Tx)
rMATS_A5SS_ISA_L
```

### rMATS_A5SS_DEXSeq vs ISA

```{r}
DEXSeq_rMATS_A5SS_ISAall <- findOverlap3(DEXSeq_rMATS_A5SS)
DEXSeq_rMATS_A5SS_ISA <- DEXSeq_rMATS_A5SS_ISAall[[1]]
DEXSeq_rMATS_A5SS_transcripts <- DEXSeq_rMATS_A5SS_ISAall[[2]]
DEXSeq_rMATS_A5SS_ISA_L <- length(unique(unlist(strsplit(as.character(
                                   DEXSeq_rMATS_A5SS_ISA$ISA_overlap), ","))))
length(ISA$transcript)
length(DEXSeq_rMATS_A5SS_transcripts)
DEXSeq_rMATS_A5SS_ISA_L
```

### rMATS_A3SS vs DEXSeq

How many exons overlap

```{r}
DEXSeq_rMATS_A3SS <- findOverlap2(rMATS_A5SSexon, 'rMATS_A3SS')
DEXSeq_rMATS_A3SS_L <- length(unique(DEXSeq_rMATS_A3SS[,'rMATS_A3SS_exon_long']))
rMATS_A3SS_L <- dim(unique(rMATS_A3SS[,c("Chr", "exon_long")]))[1]

rMATS_A3SS_L
DEXSeq_rMATS_A3SS_L
```

Relationship in results

```{r}
plot(DEXSeq_rMATS_A3SS$DEXSeq_log2fold_treatment_control,
     DEXSeq_rMATS_A3SS$rMATS_A3SS_IncLevelDifference,
     xlab="DEXSeq log2FC", ylab="rMATS IncLevelDifference", main="A3SS")
```

### rMATS_A3SS vs ISA

```{r}
rMATS_A3SS_Tx <- rMATS2Tx(rMATS_A3SSexon)
rMATS_A3SS_ISA_L <- length(intersect(rMATS_A3SS_Tx, ISA$transcript))

length(ISA$transcript)
length(rMATS_A3SS_Tx)
rMATS_A3SS_ISA_L
```

### rMATS_A3SS_DEXSeq vs ISA

```{r}
DEXSeq_rMATS_A3SS_ISAall <- findOverlap3(DEXSeq_rMATS_A3SS)
DEXSeq_rMATS_A3SS_ISA <- DEXSeq_rMATS_A3SS_ISAall[[1]]
DEXSeq_rMATS_A3SS_transcripts <- DEXSeq_rMATS_A3SS_ISAall[[2]]
DEXSeq_rMATS_A3SS_ISA_L <- length(unique(unlist(strsplit(as.character(
                                   DEXSeq_rMATS_A3SS_ISA$ISA_overlap), ","))))
length(ISA$transcript)
length(DEXSeq_rMATS_A3SS_transcripts)
DEXSeq_rMATS_A3SS_ISA_L
```

### rMATS_MXE vs DEXSeq

How many exons overlap 

```{r}
# exon 1
DEXSeq_rMATS_1_MXE <- findOverlap2(rMATS_MXE_1exon, 'rMATS_MXE')
DEXSeq_rMATS_2_MXE <- findOverlap2(rMATS_MXE_2exon, 'rMATS_MXE')
DEXSeq_rMATS_1_MXE$DEXSeq_transcripts <- NULL
DEXSeq_rMATS_1_MXE$rMATS_MXE_start <- NULL
DEXSeq_rMATS_1_MXE$rMATS_MXE_end <- NULL
DEXSeq_rMATS_2_MXE$rMATS_MXE_start <- NULL
DEXSeq_rMATS_2_MXE$rMATS_MXE_end <- NULL
# merge exon 1 and 2
DEXSeq_rMATS_MXE <- merge(DEXSeq_rMATS_1_MXE, DEXSeq_rMATS_2_MXE, all=T)
DEXSeq_rMATS_MXE <- DEXSeq_rMATS_MXE[order(DEXSeq_rMATS_MXE$DEXSeq_chr,
                                             DEXSeq_rMATS_MXE$DEXSeq_start,
                                             DEXSeq_rMATS_MXE$DEXSeq_end),]
DEXSeq_rMATS_MXE_L <- dim(unique(DEXSeq_rMATS_MXE[,c('rMATS_MXE_exon_1', 'rMATS_MXE_exon_2')]))[1]
rMATS_MXE_L <- dim(unique(rMATS_MXE[,c("Chr", "exon_1", "exon_2")]))[1]

rMATS_MXE_L
DEXSeq_rMATS_MXE_L 
```

### rMATS_MXE vs ISA

```{r}
rMATS_MXE_1_Tx <- rMATS2Tx(rMATS_MXE_1exon)
rMATS_MXE_2_Tx <- rMATS2Tx(rMATS_MXE_2exon)
rMATS_MXE_Tx <- unique(c(rMATS_MXE_1_Tx, rMATS_MXE_2_Tx))
rMATS_MXE_ISA_L <- length(intersect(rMATS_MXE_Tx, ISA$transcript))

length(ISA$transcript)
length(rMATS_MXE_Tx)
rMATS_MXE_ISA_L
```

### rMATS_MXE_DEXSeq vs ISA

```{r}
DEXSeq_rMATS_MXE_ISAall <- findOverlap3(DEXSeq_rMATS_MXE)
DEXSeq_rMATS_MXE_ISA <- DEXSeq_rMATS_MXE_ISAall[[1]]
DEXSeq_rMATS_MXE_transcripts <- DEXSeq_rMATS_MXE_ISAall[[2]]
DEXSeq_rMATS_MXE_ISA_L <- length(unique(unlist(strsplit(as.character(
                                   DEXSeq_rMATS_MXE_ISA$ISA_overlap), ","))))
length(ISA$transcript)
length(DEXSeq_rMATS_MXE_transcripts)
DEXSeq_rMATS_MXE_ISA_L
```


### rMATS_RI vs DEXSeq

How many exons overlap

```{r}
DEXSeq_rMATS_RI <- findOverlap2(rMATS_RIexon, 'rMATS_RI')
DEXSeq_rMATS_RI_L <- length(unique(DEXSeq_rMATS_RI[,'rMATS_RI_exon_ir']))
rMATS_RI_L <- dim(unique(rMATS_RI[,c("Chr", "exon_ir")]))[1]

rMATS_RI_L
DEXSeq_rMATS_RI_L
```

There is no exact match in the start and end.

```{r}
sum(DEXSeq_rMATS_RI$DEXSeq_start==DEXSeq_rMATS_RI$rMATS_RI_start &
    DEXSeq_rMATS_RI$DEXSeq_end==DEXSeq_rMATS_RI$rMATS_RI_end)
which(DEXSeq_rMATS_RI$DEXSeq_end==DEXSeq_rMATS_RI$rMATS_RI_end)
which(DEXSeq_rMATS_RI$DEXSeq_start==DEXSeq_rMATS_RI$rMATS_RI_start)
```

Relationship in results

```{r}
pdf('results/figures/RI_cor.pdf', width = 5, height = 4)
   plot(DEXSeq_rMATS_RI$DEXSeq_log2fold_treatment_control,
        DEXSeq_rMATS_RI$rMATS_RI_IncLevelDifference,
        xlab="DEXSeq log2FC", ylab="RI rMATS IncLevelDifference",
        cex=1.1, pch=16)
   abline(lm(rMATS_RI_IncLevelDifference~DEXSeq_log2fold_treatment_control,
             data=DEXSeq_rMATS_RI), lty=2)
   legend('topleft', legend = paste('r = ',
          round(cor(DEXSeq_rMATS_RI$DEXSeq_log2fold_treatment_control,
                    DEXSeq_rMATS_RI$rMATS_RI_IncLevelDifference), 2), sep = ""),
          bty='n', bg='white', cex= 1)
dev.off()

cor.test(DEXSeq_rMATS_RI$DEXSeq_log2fold_treatment_control,
                 DEXSeq_rMATS_RI$rMATS_RI_IncLevelDifference)
```

### rMATS_RI vs ISA

```{r}
rMATS_RI_Tx <- rMATS2Tx(rMATS_RIexon)
rMATS_RI_ISA_L <- length(intersect(rMATS_RI_Tx, ISA$transcript))

length(ISA$transcript)
length(rMATS_RI_Tx)
rMATS_RI_ISA_L
```

### rMATS_RI_DEXSeq vs ISA

```{r}
DEXSeq_rMATS_RI_ISAall <- findOverlap3(DEXSeq_rMATS_RI)
DEXSeq_rMATS_RI_ISA <- DEXSeq_rMATS_RI_ISAall[[1]]
DEXSeq_rMATS_RI_transcripts <- DEXSeq_rMATS_RI_ISAall[[2]]
DEXSeq_rMATS_RI_ISA_L <- length(unique(unlist(strsplit(as.character(
                                   DEXSeq_rMATS_RI_ISA$ISA_overlap), ","))))-2
length(DEXSeq_rMATS_RI_transcripts)
DEXSeq_rMATS_RI_ISA_L
```

## Venn diagramns of exon overlap

Get DEXSeq_ISA overlap

```{r}
dexseq_sing_transcrips <- dexseq_sing
colnames(dexseq_sing_transcrips)[24] <- 'DEXSeq_transcripts'
DEXSeq_ISAall <- findOverlap3(dexseq_sing_transcrips)
DEXSeq_ISA_L <- length(unique(DEXSeq_ISAall[[1]]$ISA_overlap))
```

Plot

```{r, message=F, error=F}
colors <- brewer.pal(3, 'Accent')

DEXSeq_L <- dim(dexseq_sing)[1]

ISA_L <- length(unique(ISA$transcript))

rMATS_L <- sum(rMATS_A3SS_L, rMATS_A5SS_L, rMATS_SE_L, rMATS_RI_L, rMATS_MXE_L)

DEXSeq_rMATS_L <- sum(DEXSeq_rMATS_A3SS_L, DEXSeq_rMATS_A5SS_L,
                      DEXSeq_rMATS_SE_L, DEXSeq_rMATS_RI_L,
                      DEXSeq_rMATS_MXE_L) 
rMATS_ISA_L <- sum(rMATS_A3SS_ISA_L, rMATS_A5SS_ISA_L,
                     rMATS_SE_ISA_L, rMATS_RI_ISA_L,
                     rMATS_MXE_ISA_L)
DEXSeq_rMATS_ISA_L <- sum(DEXSeq_rMATS_A3SS_ISA_L, DEXSeq_rMATS_A5SS_ISA_L,
                      DEXSeq_rMATS_SE_ISA_L, DEXSeq_rMATS_RI_ISA_L,
                      DEXSeq_rMATS_MXE_ISA_L)
grid.newpage()
draw.triple.venn(area1=DEXSeq_L, area2=rMATS_L, area3=ISA_L,
      			  n12=DEXSeq_rMATS_L, n23=rMATS_ISA_L,
      			  n13=DEXSeq_ISA_L, n123=DEXSeq_rMATS_ISA_L,
      			  category=c("DEXSeq", "rMATS", "ISA"),
      			  fill=colors)
```

```{r, echo=F}
pdf(paste(dir.fig, 'rMATS_DEXSeq_ISA_exon_overlap.pdf', sep = "/"),
    width = 5, height = 5)
grid.newpage()
draw.triple.venn(area1=DEXSeq_L, area2=rMATS_L, area3=ISA_L,
      			  n12=DEXSeq_rMATS_L, n23=rMATS_ISA_L,
      			  n13=DEXSeq_ISA_L, n123=DEXSeq_rMATS_ISA_L,
      			  category=c("DEXSeq", "rMATS", "ISA"),
      			  fill=colors)
dev.off()
```

## Write exon overlaps

```{r}
dir <- 'results/tables/splicing_overlap_exon/'
dir.create(dir, showWarnings = F)

Jtype <- c("A3SS", "A5SS", "SE", "RI", "MXE")
for (i in Jtype){
   dname <- paste('DEXSeq_rMATS_', i, '_ISA', sep = "")
   write.table(get(dname), paste(dir, dname,  '_overlap_exon.csv', sep=""),
            row.names = F, sep = '\t')
}
```

## Find gene overlap

### Create gene lists

DEXSeq

```{r}
dexseq_fdr <- 0.01
dexseqAll <- read.table('results/tables/DEXSeq/DEXSeq_results_geneFDR.csv',
                        sep = "\t", header = T, stringsAsFactors=F)
dexseqSign <- dexseqAll$gene[which(dexseqAll$gene_level_FDR < dexseq_fdr)]
dexseq_genes <- unique(grep('ENSG', unlist(strsplit(dexseqSign, "[_]|[+]")), value	= T))
length(dexseq_genes)
```

rMATS

```{r}
rmats_fdr <- 0.05
Jtype <- c("A3SS", "A5SS", "SE", "RI", "MXE")
rmats <- c()
for (i in Jtype){
   #i <- "SE"
   # d <- read.table(paste('results/tables/rMATS/rMATS_sign_', i, '.csv', sep = ""),
   #                 header = T, sep = "\t", stringsAsFactors=F)
   # rmats <- unique(c(rmats, d$GeneID))
   
   # to include unfiltered rMATS results
   d <- read.table(paste('results/tables/rMATS/rMATS_all_', i, '.csv', sep = ""),
                   header = T, sep = "\t", stringsAsFactors=F)
   dsign <- as.character(d$GeneID[which(d$FDR < rmats_fdr)])
   rmats <- unique(c(rmats, dsign))
}
length(rmats)
```

ISA

```{r}
ISA_fdr <- 0.05
ISAgenes <- unique(ISA$gene_id[ISA$padj<=ISA_fdr])
length(ISAgenes)
```

### Overlap

Plot

```{r}
vennPlot <- venn.diagram(list(rmats, dexseq_genes, ISAgenes),
                         NULL,
                         fill = brewer.pal(3, 'Accent'),
                         category.names = c('rMATS', 'DEXSeq', 'ISA'))
grid.draw(vennPlot)
```

Save to file

```{r}
venn.diagram(list(rmats, dexseq_genes, ISAgenes),
             paste(dir.fig, 'rMATS_DEXSeq_ISA_gene_overlap.pdf', sep = "/"),
             fill = brewer.pal(3, 'Accent'),
             category.names = c('rMATS', 'DEXSeq', 'ISA'))
```

```{bash clean_logs, include=FALSE}
rm VennDiagram*.log results/figures/rMATS_DEXSeq_ISA_overlap/*.log
```

What proportion is it?

```{r}
rmats_dexseq <- intersect(rmats, dexseq_genes)
rmats_isa <- intersect(rmats, ISAgenes)
isa_dexseq <- intersect(ISAgenes, dexseq_genes)
rmats_dexseq_isa <- intersect(rmats_dexseq, ISAgenes)
rmats_dexseq_isa_atleast2 <- unique(c(rmats_dexseq, rmats_isa, isa_dexseq))

length(rmats_dexseq)/length(rmats)
length(rmats_isa)/length(rmats)
length(isa_dexseq)/length(ISAgenes)
length(rmats_dexseq_isa)/length(rmats)
length(rmats_dexseq_isa_atleast2)
```

### Annotate the overlap genes

In all 3 methods

```{r, message=FALSE, error=FALSE, eval=TRUE}
rmats_dexseq_isa_renamed <- gsub("\\..*","", rmats_dexseq_isa)

rmats_dexseq_isa_annotAll <- queryMany(unique(rmats_dexseq_isa_renamed),
                      scopes='ensembl.gene',
                      fields='entrezgene,symbol,name,summary', # pathway.kegg.id,go
                      return.as='DataFrame',
                      species='human',
                      returnall=FALSE)
rmats_dexseq_isa_annot <- as.data.frame(rmats_dexseq_isa_annotAll[,c('query', 'entrezgene','symbol', 'name', 'summary')])
colnames(rmats_dexseq_isa_annot)[c(1,2)] <- c('gene', 'entrez')
```

In at least 2 methods

```{r, message=FALSE, error=FALSE, eval=TRUE}
rmats_dexseq_isa_atleast2_renamed <- gsub("\\..*","", rmats_dexseq_isa_atleast2)
rmats_dexseq_isa_atleast2_annotAll <- queryMany(unique(rmats_dexseq_isa_atleast2_renamed),
                      scopes='ensembl.gene',
                      fields='entrezgene,symbol,name,summary', # pathway.kegg.id,go
                      return.as='DataFrame',
                      species='human',
                      returnall=FALSE)
rmats_dexseq_isa_atleast2_annot <- as.data.frame(rmats_dexseq_isa_atleast2_annotAll[,c('query', 'entrezgene','symbol', 'name', 'summary')])
colnames(rmats_dexseq_isa_atleast2_annot)[c(1,2)] <- c('gene', 'entrez')
```

### Write results

Genes that occur in at least two tests

```{r}
write.table(rmats_dexseq_isa_atleast2_annot,
            paste(dir.table, 'rMATS_DEXSeq_ISA_gene_atleast2overlap.csv', sep = "/"),
            quote = T,
            row.names = F,
            sep = '\t')
```

Genes that occur in  all three  tests

```{r}
write.table(rmats_dexseq_isa_annot,
            paste(dir.table, 'rMATS_DEXSeq_ISA_gene_all3overlap.csv', sep = "/"),
            quote = T,
            row.names = F,
            sep = '\t')
```
