---
title: "Myocardin splicing: overlap between methods"
author:
   name: "Dmytro Kryvokhyzha"
   email: dmytro.kryvokhyzha@med.lu.se
   affiliation: LUDC Bioinformatics Unit
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, eval=TRUE)
knitr::opts_knit$set(root.dir = '../')
``` 

## Dependencies

```{r, message=FALSE, eval=TRUE}
library(VennDiagram)
library(RColorBrewer)
library(mygene)
```

## Data

DEXSeq

```{r}
dexseq_fdr <- 0.01
dexseqAll <- read.table('results/tables/DEXSeq/DEXSeq_results_geneFDR.csv',
                        sep = "\t", header = T, stringsAsFactors=F)
dexseqSign <- dexseqAll$gene[which(dexseqAll$gene_level_FDR < dexseq_fdr)]
dexseq <- unique(grep('ENSG', unlist(strsplit(dexseqSign, "[_]|[+]")), value	= T))
length(dexseq)
```

rMATS

```{r}
rmats_fdr <- 0.05
Jtype <- c("A3SS", "A5SS", "SE", "RI", "MXE")
rmats <- c()
for (i in Jtype){
   #i <- "SE"
   # d <- read.table(paste('results/tables/rMATS_sign/rMATS_sign_', i, '.csv', sep = ""),
   #                 header = T, sep = "\t", stringsAsFactors=F)
   # rmats <- unique(c(rmats, d$GeneID))
   
   # to include unfiltered rMATS results
   d <- read.table(paste('results/tables/rMATS/rMATS_all_', i, '.csv', sep = ""),
                   header = T, sep = "\t", stringsAsFactors=F)
   dsign <- as.character(d$GeneID[which(d$FDR < rmats_fdr)])
   rmats <- unique(c(rmats, dsign))
}
length(rmats)
```

SUPPA isoforms

```{r}
suppa_fdr <- 0.05
suppa_isoformAll <- read.table('results/tables/SUPPA/SUPPA_all.csv',
                            sep = "\t", header = T, stringsAsFactors=F)
suppa_isoform_sing <- which(suppa_isoformAll$p.value < suppa_fdr)
suppa_isoform <- na.omit(unique(as.character(suppa_isoformAll$gene[suppa_isoform_sing])))
length(suppa_isoform)
```

## Overlap

Plot

```{r}
vennPlot <- venn.diagram(list(rmats, dexseq, suppa_isoform),
                         NULL,
                         fill = brewer.pal(3, 'Accent'),
                         category.names = c('rMATS', 'DEXSeq', 'SUPPA isoform'))
grid.draw(vennPlot)
```

Save to file

```{r}
venn.diagram(list(rmats, dexseq, suppa_isoform),
             'results/figures/rmats_dexseq_suppa.pdf',
             fill = brewer.pal(3, 'Accent'),
             category.names = c('rMATS', 'DEXSeq', 'SUPPA isoform'))
```

```{bash clean_logs, include=FALSE}
rm VennDiagram*.log results/figures/*.log
```

What proportion is it?

```{r}
rmats_dexseq <- intersect(rmats, dexseq)
rmats_suppa <- intersect(rmats, suppa_isoform)
suppa_dexseq <- intersect(suppa_isoform, dexseq)
rmats_dexseq_suppa <- intersect(rmats_dexseq, suppa_isoform)
rmats_dexseq_suppa_togather <- unique(c(rmats_dexseq, rmats_suppa, suppa_dexseq))

length(rmats_dexseq)/length(rmats)
length(rmats_suppa)/length(rmats)
length(suppa_dexseq)/length(suppa_isoform)
length(rmats_dexseq_suppa)/length(rmats)
length(rmats_dexseq_suppa_togather)
```

## Annotate the overlap genes

```{r, message=FALSE, error=FALSE, eval=TRUE}
gene_renamed <- gsub("\\..*","", rmats_dexseq_suppa_togather)
annotAll <- queryMany(unique(gene_renamed),
                      scopes='ensembl.gene',
                      fields='entrezgene,symbol,name,summary', # pathway.kegg.id,go
                      return.as='DataFrame',
                      species='human',
                      returnall=FALSE)
rmats_dexseq_suppa_togather_annot <- as.data.frame(annotAll[,c('query', 'entrezgene','symbol', 'name', 'summary')])
colnames(rmats_dexseq_suppa_togather_annot)[c(1,2)] <- c('gene', 'entrez')
```

## Control genes

```{r}
interesting_genes <- c('MYOCD', 'MYH11', 'MCAM', 'SYNM', 'RRAS', 'RBPMS', 'RBPMS2', 'KCNMB1',
                       'ACTN1', 'SMTN', 'NFIC', 'BICD2', 'TEAD1', 'TPM1', 'TPM2',
                       'PDLIM7', 'FHL1', 'CSR2', 'TGFB11', 'PKD1', 'MYL6', 'PDLIM5',
                       'FLNA', 'MBNL1', 'ZFR', 'FNBP1', 'SORBS1', 'PIP5K1C',
                       'TNRC18', 'SVIL', 'KLHL42', 'LARGE1', 'MICAL3')
```

These genes in at least two analyses:

```{r}
interesting_genes_rows <- grep(paste(interesting_genes,collapse="|"),
                               rmats_dexseq_suppa_togather_annot$symbol)
rmats_dexseq_suppa_togather_int <- rmats_dexseq_suppa_togather_annot[
                                                         interesting_genes_rows,]
knitr::kable(head(rmats_dexseq_suppa_togather_int[,c(1:4)]))
```

These genes in all three analyses:

```{r}
rmats_dexseq_suppa_id <- gsub("\\..*","", rmats_dexseq_suppa)
allTheeRows <- which(rmats_dexseq_suppa_togather_int$gene %in% rmats_dexseq_suppa_id)
rmats_dexseq_suppa_int <- rmats_dexseq_suppa_togather_int[allTheeRows,]
knitr::kable(head(rmats_dexseq_suppa_int[,c(1:4)]))
```

## Write results

Genes that occur in at least two tests

```{r}
write.table(rmats_dexseq_suppa_togather_int,
            'results/tables/rMATS_DEXSeq_SUPPA_atleast2overlap.csv',
            quote = T,
            row.names = F,
            sep = '\t')
```

Genes that occur in  all three  tests

```{r}
write.table(rmats_dexseq_suppa_int,
            'results/tables/rMATS_DEXSeq_SUPPA_overlap.csv',
            quote = T,
            row.names = F,
            sep = '\t')
```
